cli/src/commands/rebase.rs

```diff
@@ -356,96 +356,47 @@ fn rebase_revision(
+    if new_parents.contains(&to_rebase_commit) {
         return Err(user_error(format!(
             "Cannot rebase {} onto itself",
-            short_commit_hash(old_commit.id()),
+            short_commit_hash(to_rebase_commit.id()),
         )));
     }
 
-    let children_expression = RevsetExpression::commit(old_commit.id().clone()).children();
-    let child_commits: Vec<_> = children_expression
-        .evaluate_programmatic(workspace_command.repo().as_ref())
-        .unwrap()
-        .iter()
-        .commits(workspace_command.repo().store())
-        .try_collect()?;
-    // Currently, immutable commits are defined so that a child of a rewriteable
-    // commit is always rewriteable.
-    debug_assert!(workspace_command
-        .check_rewritable(child_commits.iter().ids())
-        .is_ok());
-
-    // First, rebase the children of `old_commit`.
     let mut tx = workspace_command.start_transaction();
     let mut rebased_commit_ids = HashMap::new();
-    for child_commit in child_commits {
-        let new_child_parent_ids: Vec<CommitId> = child_commit
-            .parents()
-            .iter()
-            .flat_map(|c| {
-                if c == &old_commit {
-                    old_commit.parent_ids().to_vec()
-                } else {
-                    [c.id().clone()].to_vec()
-                }
-            })
-            .collect();
-
-        // Some of the new parents may be ancestors of others as in
-        // `test_rebase_single_revision`.
-        let new_child_parents_expression = RevsetExpression::commits(new_child_parent_ids.clone())
-            .minus(
-                &RevsetExpression::commits(new_child_parent_ids.clone())
-                    .parents()
-                    .ancestors(),
-            );
-        let new_child_parents = new_child_parents_expression
-            .evaluate_programmatic(tx.base_repo().as_ref())
-            .unwrap()
-            .iter()
-            .collect_vec();
 
-        rebased_commit_ids.insert(
-            child_commit.id().clone(),
-            rebase_commit(settings, tx.mut_repo(), child_commit, new_child_parents)?
-                .id()
-                .clone(),
-        );
-    }
-    // Now, rebase the descendants of the children.
+    // First, rebase the descendants of `old_commit`.
     // TODO(ilyagr): Consider making it possible for these descendants to become
     // emptied, like --skip_empty. This would require writing careful tests.
-    rebased_commit_ids.extend(tx.mut_repo().rebase_descendants_return_map(settings)?);
+    tx.mut_repo().transform_descendants(
+        settings,
+        vec![to_rebase_commit.id().clone()],
+        |mut rewriter| {
+            let old_commit = rewriter.old_commit();
+            let old_commit_id = old_commit.id().clone();
+
+            // Replace references to `old_commit` with its parents.
+            if old_commit.parent_ids().contains(to_rebase_commit.id()) {
```

Comment:
nit: this optimization is already part of `replace_parent()`, so you can remove it from here

Reply:
Oh, but since we're keeping the `simplify_ancestor_merge()` for now, I suppose it'll have to stay

Reply:
Updated to remove in the subsequent commit.